File: typescript/src/lesson.ts
import { AIMessage } from "@langchain/core/messages";
import type { RunnableConfig } from "@langchain/core/runnables";
import { tool } from "@langchain/core/tools";
import { FakeListChatModel } from "@langchain/core/utils/testing";
import { createDeepAgent, StateBackend } from "deepagents";
import { z } from "zod";
import { createDataScopeGate, createScopeLedger, loadScopePolicy, type ScopePolicy } from "./gate.js";
import { scoreNecessity } from "./necessity.js";

/**
 * A tiny fictional agent for the data scope lesson. No model is paid for:
 * the "model" is a script that asks for two tools, one permitted and one not.
 * The permitted tool returns two records, two from the right case and one
 * from another, and one of them carries a forbidden field.
 */

export const LESSON_AGENT = "example-investigator";
export const LESSON_POLICY_PATH = "../policies/example-scope.yaml";
export const LESSON_INCIDENT = "inc-fictional-0001";
const EMPTY = z.strictObject({});

export const LESSON_SCI_TOOL = "get_support_message_preview";

export function createDataScopeLessonAgent(
  model: FakeListChatModel,
  options: { onBlock?: ScopePolicy["defaults"]["on_block"]; policyPath?: string; sciProbe?: boolean } = {},
) {
  const loaded = loadScopePolicy(options.policyPath ?? LESSON_POLICY_PATH);
  // Lesson-only overrides, applied to a copy of the policy: the owner's on_block
  // choice, and (for the incident lesson) one extra permitted source at the
  // most sensitive tier, so a strip on it can be watched becoming an incident.
  const block = loaded.agents[LESSON_AGENT];
  const policy: ScopePolicy = {
    ...loaded,
    agents: { ...loaded.agents, [LESSON_AGENT]: {
      ...block,
      ...(options.onBlock ? { on_block: options.onBlock } : {}),
      permit: options.sciProbe ? [...block.permit, { tool: LESSON_SCI_TOOL, tier: "SCI" }] : block.permit,
    } },
  };
  const ledger = createScopeLedger(policy);
  const ran = { permittedTool: 0, forbiddenTool: 0, sciTool: 0 };

  const platformAggregate = tool(async () => {
    ran.permittedTool++;
    return JSON.stringify({ version: 1, records: [
      { ref: "snapshot:fictional:signals", incidentId: LESSON_INCIDENT, kind: "signals", value: { fictional: true, signals: ["delivery_failure"] }, email_address: "fictional@example.invalid" },
      { ref: "snapshot:fictional:heartbeat", incidentId: LESSON_INCIDENT, kind: "heartbeat", value: { fictional: true, heartbeat: "fresh" } },
      { ref: "snapshot:fictional-other:platform", incidentId: "inc-fictional-9999", kind: "platform", value: { fictional: true, heartbeat: "stale" } },
    ] });
  }, { name: "get_incident_snapshot", description: "Read the fictional incident snapshot.", schema: EMPTY });

  const messagePreview = tool(async () => {
    ran.sciTool++;
    return JSON.stringify({ version: 1, records: [
      { ref: "message:fictional:1", incidentId: LESSON_INCIDENT, kind: "support-message", subject: "fictional subject", message_text: "fictional body that must never reach the model" },
    ] });
  }, { name: LESSON_SCI_TOOL, description: "Lesson only: a sensitive-tier source that returns a forbidden field.", schema: EMPTY });

  const householdRecords = tool(async () => {
    ran.forbiddenTool++;
    return JSON.stringify({ fictional: true, households: [] });
  }, { name: "get_household_records", description: "A tool this agent is not permitted to use.", schema: EMPTY });

  const agent = createDeepAgent({
    model,
    name: LESSON_AGENT,
    backend: new StateBackend(),
    subagents: [],
    tools: options.sciProbe ? [platformAggregate, householdRecords, messagePreview] : [platformAggregate, householdRecords],
    systemPrompt: "Fictional lesson agent. Read the incident snapshot, then report.",
    middleware: [createDataScopeGate({
      policy, agent: LESSON_AGENT,
      context: { incident: { id: LESSON_INCIDENT } },
      onDecision: ledger.record,
    })],
  });
  return { agent, policy, ledger, ran };
}

/** The script the fake model follows: call the tools, then answer. */
export function scriptedLessonModel(options: { sciProbe?: boolean } = {}) {
  const model = new FakeListChatModel({ responses: ["unused"] });
  model.getName = () => "ChatOpenAI";
  model.bindTools = () => model;
  let calls = 0;
  model._generate = async () => {
    const message = ++calls === 1
      ? new AIMessage({ content: "", tool_calls: [
          { name: "get_incident_snapshot", args: {}, id: "lesson-read-1", type: "tool_call" },
          { name: "get_household_records", args: {}, id: "lesson-read-2", type: "tool_call" },
          ...(options.sciProbe ? [{ name: LESSON_SCI_TOOL, args: {}, id: "lesson-read-3", type: "tool_call" as const }] : []),
        ] })
      : new AIMessage(JSON.stringify({ finding: "Fictional report: one delivery failure signal; household records were not available.", evidenceRefs: ["snapshot:fictional:signals"] }));
    return { generations: [{ text: String(message.content), message }] };
  };
  return model;
}

export async function runDataScopeLesson(
  session: ReturnType<typeof createDataScopeLessonAgent>,
  traceConfig: Pick<RunnableConfig, "callbacks" | "runId" | "runName" | "tags"> = {},
) {
  const result = await session.agent.invoke(
    { messages: [{ role: "user", content: "Run the fictional data scope lesson." }] },
    { ...traceConfig, recursionLimit: 8, signal: AbortSignal.timeout(30_000) },
  );
  const toolMessages = result.messages.filter((message) => message.type === "tool");
  const finalText = String(result.messages.at(-1)?.content ?? "");
  let citedRefs: string[] = [];
  try {
    const parsed = JSON.parse(finalText) as { evidenceRefs?: unknown };
    if (Array.isArray(parsed.evidenceRefs)) citedRefs = parsed.evidenceRefs.filter((ref): ref is string => typeof ref === "string");
  } catch { /* a run that ends without a report cites nothing */ }
  const necessity = scoreNecessity(session.ledger.decisions, new Map([[LESSON_AGENT, citedRefs]]));
  return {
    finalText,
    citedRefs,
    necessity,
    toolMessages,
    decisions: session.ledger.decisions,
    summary: session.ledger.summary(),
    mark: session.ledger.mark(),
    ran: session.ran,
  };
}

File: python/agent_data_scope/lesson.py
"""A tiny fictional agent for the data scope lesson. No model is paid for: the
"model" is a script that asks for two tools, one permitted and one not. The
permitted tool returns records from the right incident and one from another,
and one of them carries a forbidden field.
"""

from __future__ import annotations

import json
import os
from dataclasses import dataclass, field
from typing import Any, Iterator, Sequence

from deepagents import create_deep_agent
from deepagents.backends import StateBackend
from langchain_core.language_models.fake_chat_models import GenericFakeChatModel
from langchain_core.messages import AIMessage, ToolMessage
from langchain_core.tools import tool

from .gate import GateDecision, LedgerSummary, ScopeLedger, create_data_scope_gate, create_scope_ledger
from .necessity import NecessityScore, score_necessity
from .policy import OnBlock, ScopePolicy, load_scope_policy

LESSON_AGENT = "example-investigator"
LESSON_INCIDENT = "inc-fictional-0001"
LESSON_SCI_TOOL = "get_support_message_preview"
LESSON_POLICY_PATH = os.path.join(os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))), "policies", "example-scope.yaml")
"""``<repo>/policies/example-scope.yaml``, shared with the TypeScript port."""


class ScriptedChatModel(GenericFakeChatModel):
    """A fake chat model that plays back scripted AIMessages and accepts tool binding."""

    def bind_tools(self, tools: Sequence[Any], **kwargs: Any) -> "ScriptedChatModel":
        return self

    @property
    def _llm_type(self) -> str:
        return "scripted-lesson-model"


def scripted_lesson_model(sci_probe: bool = False) -> ScriptedChatModel:
    """The script the fake model follows: call the tools, then answer."""
    calls = [
        {"name": "get_incident_snapshot", "args": {}, "id": "lesson-read-1", "type": "tool_call"},
        {"name": "get_household_records", "args": {}, "id": "lesson-read-2", "type": "tool_call"},
    ]
    if sci_probe:
        calls.append({"name": LESSON_SCI_TOOL, "args": {}, "id": "lesson-read-3", "type": "tool_call"})
    final = json.dumps({"finding": "Fictional report: one delivery failure signal; household records were not available.", "evidenceRefs": ["snapshot:fictional:signals"]})

    def script() -> Iterator[AIMessage]:
        yield AIMessage(content="", tool_calls=calls)
        yield AIMessage(content=final)

    return ScriptedChatModel(messages=script())


@dataclass
class LessonSession:
    agent: Any
    policy: ScopePolicy
    ledger: ScopeLedger
    ran: dict[str, int] = field(default_factory=lambda: {"permitted_tool": 0, "forbidden_tool": 0, "sci_tool": 0})


def create_lesson_agent(model: Any, on_block: OnBlock | None = None, policy_path: str | None = None, sci_probe: bool = False) -> LessonSession:
    loaded = load_scope_policy(policy_path or LESSON_POLICY_PATH)
    # Lesson-only overrides, applied to a copy of the policy: the owner's on_block
    # choice, and (for the incident lesson) one extra permitted source at the
    # most sensitive tier, so a strip on it can be watched becoming an incident.
    copy = loaded.model_dump()
    block = copy["agents"][LESSON_AGENT]
    if on_block:
        block["on_block"] = on_block
    if sci_probe:
        block["permit"] = [*block["permit"], {"tool": LESSON_SCI_TOOL, "tier": "SCI"}]
    policy = ScopePolicy.model_validate(copy)
    ledger = create_scope_ledger(policy)
    ran = {"permitted_tool": 0, "forbidden_tool": 0, "sci_tool": 0}

    @tool
    def get_incident_snapshot() -> str:
        """Read fixed fictional incident signals and totals."""
        ran["permitted_tool"] += 1
        return json.dumps(
            {
                "version": 1,
                "records": [
                    {"ref": "snapshot:fictional:signals", "incidentId": LESSON_INCIDENT, "kind": "snapshot-signals", "value": {"fictional": True, "signals": ["delivery_failure"]}, "email_address": "fictional@example.invalid"},
                    {"ref": "snapshot:fictional:heartbeat", "incidentId": LESSON_INCIDENT, "kind": "snapshot-heartbeat", "value": {"fictional": True, "heartbeat": "fresh"}},
                    {"ref": "snapshot:fictional-other:platform", "incidentId": "inc-fictional-9999", "kind": "snapshot-platform", "value": {"fictional": True, "heartbeat": "stale"}},
                ],
            }
        )

    @tool
    def get_household_records() -> str:
        """A tool this agent is not permitted to use."""
        ran["forbidden_tool"] += 1
        return json.dumps({"fictional": True, "households": []})

    @tool(LESSON_SCI_TOOL)
    def get_support_message_preview() -> str:
        """Lesson only: a sensitive-tier source that returns a forbidden field."""
        ran["sci_tool"] += 1
        return json.dumps(
            {
                "version": 1,
                "records": [
                    {"ref": "message:fictional:1", "incidentId": LESSON_INCIDENT, "kind": "support-message", "subject": "fictional subject", "message_text": "fictional body that must never reach the model"},
                ],
            }
        )

    tools = [get_incident_snapshot, get_household_records] + ([get_support_message_preview] if sci_probe else [])
    agent = create_deep_agent(
        model=model,
        tools=tools,
        name=LESSON_AGENT,
        backend=StateBackend(),
        subagents=[],
        system_prompt="Fictional lesson agent. Read the incident snapshot, then report.",
        middleware=[create_data_scope_gate(policy, LESSON_AGENT, context={"incident": {"id": LESSON_INCIDENT}}, on_decision=ledger.record)],
    )
    return LessonSession(agent=agent, policy=policy, ledger=ledger, ran=ran)


@dataclass
class LessonOutcome:
    final_text: str
    cited_refs: list[str]
    necessity: NecessityScore
    tool_messages: list[ToolMessage]
    decisions: list[GateDecision]
    summary: LedgerSummary
    mark: str
    ran: dict[str, int]


def run_lesson(session: LessonSession, config: dict[str, Any] | None = None) -> LessonOutcome:
    """Runs the lesson. ``config`` may carry callbacks, run_id, run_name and tags for tracing."""
    result = session.agent.invoke({"messages": [{"role": "user", "content": "Run the fictional data scope lesson."}]}, {**(config or {}), "recursion_limit": 8})
    messages = result["messages"]
    tool_messages = [message for message in messages if isinstance(message, ToolMessage)]
    final_text = str(messages[-1].content) if messages else ""
    cited_refs: list[str] = []
    try:
        parsed = json.loads(final_text)
        refs = parsed.get("evidenceRefs") if isinstance(parsed, dict) else None
        if isinstance(refs, list):
            cited_refs = [ref for ref in refs if isinstance(ref, str)]
    except ValueError:
        pass  # a run that ends without a report cites nothing
    necessity = score_necessity(session.ledger.decisions, {LESSON_AGENT: cited_refs})
    return LessonOutcome(
        final_text=final_text,
        cited_refs=cited_refs,
        necessity=necessity,
        tool_messages=tool_messages,
        decisions=list(session.ledger.decisions),
        summary=session.ledger.summary(),
        mark=session.ledger.mark(),
        ran=session.ran,
    )
